File: app/src/androidTest/java/ru/org/adons/cblock/datamodel/BlockListModelTest.java
package ru.org.adons.cblock.datamodel;

import android.text.format.DateUtils;

import org.junit.Before;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import javax.inject.Inject;

import ru.org.adons.cblock.model.BlockListItem;
import ru.org.adons.cblock.model.CallLogItem;
import ru.org.adons.cblock.utils.Logging;
import rx.observers.TestSubscriber;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertNotNull;

/**
 * Android test for {@link BlockListModel}
 */

public class BlockListModelTest extends BaseModelTest {

    @Inject BlockListModel blockListModel;
    private final List<BlockListItem> blockList = new ArrayList<>();

    @Before
    public void setUp() {
        // provide dependencies
        testComponent().inject(this);
        // clear all local data before test
        blockListModel.clearBlockList();
    }

    @Test
    public void addNumberToBlockList() {
        String phoneNumber1 = "123456789";
        String phoneNumber2 = "987654321";

        // check number added to block list
        addNumber(phoneNumber1);
        assertEquals(blockList.size(), 1);
        BlockListItem listItem = blockList.get(0);
        assertNotNull(listItem);
        assertEquals(listItem.getPhoneNumber(), phoneNumber1);

        // check not duplicated
        addNumber(phoneNumber1);
        assertEquals(blockList.size(), 1);
        listItem = blockList.get(0);
        assertNotNull(listItem);
        assertEquals(listItem.getPhoneNumber(), phoneNumber1);

        // check add other number
        addNumber(phoneNumber2);
        assertEquals(blockList.size(), 2);
        listItem = blockList.get(0);
        BlockListItem listItem1 = blockList.get(1);
        assertFalse(listItem.getPhoneNumber().equals(listItem1.getPhoneNumber()));

        // check not duplicated
        addNumber(phoneNumber1);
        assertEquals(blockList.size(), 2);
        addNumber(phoneNumber2);
        assertEquals(blockList.size(), 2);
    }

    private void addNumber(String phoneNumber) {
        TestSubscriber<BlockListItem> itemSubscriber = new TestSubscriber<>();
        TestSubscriber<List<BlockListItem>> listSubscriber = new TestSubscriber<>();
        CallLogItem logItem = getLogItem(phoneNumber);
        blockListModel.addNumber(logItem)
                .doOnSubscribe(Logging.actionSubscribe(this.getClass(), "addNumber"))
                .doOnUnsubscribe(Logging.actionUnsubscribe(this.getClass(), "addNumber"))
                .subscribe(itemSubscriber);
        itemSubscriber.assertNoErrors();
        itemSubscriber.assertCompleted();
        itemSubscriber.assertUnsubscribed();

        blockListModel.getBlockList()
                .doOnSubscribe(Logging.actionSubscribe(this.getClass(), "getBlockList"))
                .doOnUnsubscribe(Logging.actionUnsubscribe(this.getClass(), "getBlockList"))
                .doOnNext(this::setBlockList)
                .subscribe(listSubscriber);
        listSubscriber.assertNoErrors();
        listSubscriber.assertCompleted();
        listSubscriber.assertUnsubscribed();
    }

    private CallLogItem getLogItem(String phoneNumber) {
        return CallLogItem.builder()
                .setId(1L)
                .setPhoneNumber(phoneNumber)
                .setDate(new Date().getTime())
                .build();
    }

    private void setBlockList(List<BlockListItem> items) {
        blockList.clear();
        blockList.addAll(items);
        // print details
        for (BlockListItem item : items) {
            Logging.d("id=" + item.getId().toString());
            Logging.d("phone=" + item.getPhoneNumber());
            Logging.d("date=" + DateUtils.getRelativeTimeSpanString(item.getDate()).toString());
            Logging.d("name=" + item.getName());
        }
    }

}

File: app/src/androidTest/java/mobile/addons/cblock/data/BlockListModelTest.kt
package mobile.addons.cblock.data

import android.support.test.runner.AndroidJUnit4
import android.text.format.DateUtils
import mobile.addons.cblock.ext.log
import mobile.addons.cblock.model.BlockListModel
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import rx.Observable
import rx.observers.TestSubscriber
import java.util.*
import javax.inject.Inject

/**
 * Android test for [BlockListModel]
 */
@RunWith(AndroidJUnit4::class)
class BlockListModelTest : BaseModelTest() {

    private val phoneNumber1 = "+380123456789"
    private val phoneNumber2 = "+380987654321"
    private val name1 = "Test Name"

    @Inject lateinit var blockListModel: BlockListModel

    @Before
    fun setUp() {
        // provide dependencies
        testComponent().inject(this)
        // clear all local data before test
        blockListModel.clearBlockList()
    }

    @Test
    fun addNumberToBlockList() {
        log("check number added to block list")
        addNumber(phoneNumber1)
        //
        log("check not duplicated")
        addNumber(phoneNumber1)
        //
        log("check add other number")
        addOtherNumber()
        //
        log("check not duplicated")
        addSameTwoNumbers()
    }

    private fun addNumber(phoneNumber: String) {
        val testSubscriber = TestSubscriber<BlockListItem>()
        blockListModel.addNumber(getLogItem(phoneNumber, name1))
        blockListModel.getBlockList()
                .flatMap { Observable.from(it) }
                .doOnNext {
                    assertNotNull(it)
                    assertEquals(it.phoneNumber, phoneNumber)
                    printItemDetails(it)
                }
                .subscribe(testSubscriber)
        testSubscriber.assertValueCount(1)
        testSubscriber.assertNoErrors()
        testSubscriber.assertCompleted()
        testSubscriber.assertUnsubscribed()
    }

    private fun addOtherNumber() {
        val testSubscriber = TestSubscriber<BlockListItem>()
        blockListModel.addNumber(getLogItem(phoneNumber2, null))
        blockListModel.getBlockList()
                .doOnNext {
                    val listItem = it[0]
                    val listItem1 = it[1]
                    assertFalse(listItem.phoneNumber == listItem1.phoneNumber)
                }
                .flatMap { Observable.from(it) }
                .doOnNext { this.printItemDetails(it) }
                .subscribe(testSubscriber)
        testSubscriber.assertValueCount(2)
        testSubscriber.assertNoErrors()
        testSubscriber.assertCompleted()
        testSubscriber.assertUnsubscribed()
    }

    private fun addSameTwoNumbers() {
        val testSubscriber = TestSubscriber<BlockListItem>()
        blockListModel.addNumber(getLogItem(phoneNumber1, name1))
        blockListModel.addNumber(getLogItem(phoneNumber2, null))
        blockListModel.getBlockList()
                .flatMap { Observable.from(it) }
                .doOnNext { this.printItemDetails(it) }
                .subscribe(testSubscriber)
        testSubscriber.assertValueCount(2)
        testSubscriber.assertNoErrors()
        testSubscriber.assertCompleted()
        testSubscriber.assertUnsubscribed()
    }

    private fun getLogItem(phoneNumber: String, name: String?): CallLogItem = CallLogItem(
            1L, phoneNumber, Date().time, name, false, false)

    private fun printItemDetails(item: BlockListItem): BlockListItem = with(item) {
        log("id=$id")
        log("phone=$phoneNumber")
        log("date=" + DateUtils.getRelativeTimeSpanString(date))
        log("name=$name")
        item
    }

}
